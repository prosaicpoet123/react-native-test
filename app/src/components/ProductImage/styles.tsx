import { StyleSheet } from 'react-native'
import { colours } from '../../config/themes'

const styles = StyleSheet.create({
  productImage: {
    width: '100%',
    height: 180,
  },
  thumbnails: {
    display: 'flex',
    flexWrap: 'wrap',
    flexDirection: 'row',
    marginVertical: 8,
  },
  thumbnail: {
    height: 70,
    width: 70,
  },
  selectedThumbnail: {
    borderColor: colours.darkGreenPrimary,
    borderWidth: 1,
  },
  imageWrapper: {
    position: 'relative',
  },
  promotionsWrapper: {
    position: 'absolute',
    top: 4,
    left: 4,
    zIndex: 1,
  },
})

export default styles
