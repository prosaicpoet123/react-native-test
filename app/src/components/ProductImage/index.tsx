import React, { FC, useState } from 'react'
import { FlatList, Image, TouchableOpacity, View } from 'react-native'
import { IProduct, IProductImagesList } from '../../@Types/product'
import Promotions from '../Promotions'
import { parseImagesByResolution } from '../../utils/parseImageByResolution'
import styles from './styles'

interface IProductImage {
  id: string
  images: IProductImagesList[]
  resolution?: number
  showThumbnails?: boolean
  promotions?: IProduct['promotions']
}

type IProps = IProductImage

const ProductImage: FC<IProps> = ({
  id,
  images,
  showThumbnails,
  promotions,
  resolution = 180,
}) => {
  const [currentImage, setCurrentImage] = useState('A')
  const imagePositions = ['A', 'B', 'C', 'D', 'E', 'F', 'G']
  const imageSrc = `https://images.hollandandbarrettimages.co.uk/productimages/HB/${resolution}/${id}_${currentImage}.jpg`

  const thumbnails = showThumbnails ? parseImagesByResolution(images, 70) : []

  return (
    <View>
      <View style={styles.imageWrapper}>
        {!!promotions?.length && (
          <View style={styles.promotionsWrapper}>
            <Promotions
              promotions={promotions.map((promotion) => promotion.text)}
            />
          </View>
        )}
        <Image
          style={[styles.productImage, { height: resolution }]}
          source={{
            uri: imageSrc,
          }}
        />
      </View>
      {!!showThumbnails && (
        <View style={styles.thumbnails}>
          <FlatList
            horizontal
            data={thumbnails}
            renderItem={({ item, index }) => (
              <TouchableOpacity
                style={[
                  currentImage === imagePositions[index] &&
                    styles.selectedThumbnail,
                ]}
                onPress={() => setCurrentImage(imagePositions[index])}
                disabled={currentImage === imagePositions[index]}>
                <Image
                  style={styles.thumbnail}
                  source={{
                    uri: `https:${item?.path}`,
                  }}
                />
              </TouchableOpacity>
            )}
            keyExtractor={(item, index) => imagePositions[index]}
          />
        </View>
      )}
    </View>
  )
}

export default ProductImage
