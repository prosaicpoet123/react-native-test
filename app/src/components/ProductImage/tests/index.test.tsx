import React from 'react'
import renderer from 'react-test-renderer'
import ProductImage from '../'

const images = [
  {
    list: [
      {
        path: '//images.hollandandbarrettimages.co.uk/productimages/HB/120/048774_A.jpg',
        resolution: 120,
      },
      {
        path: '//images.hollandandbarrettimages.co.uk/productimages/HB/180/048774_A.jpg',
        resolution: 180,
      },
      {
        path: '//images.hollandandbarrettimages.co.uk/productimages/HB/252/048774_A.jpg',
        resolution: 252,
      },
    ],
    group: '1',
  },
  {
    list: [
      {
        path: '//images.hollandandbarrettimages.co.uk/productimages/HB/370/048774_A.jpg',
        resolution: 370,
      },
      {
        path: '//images.hollandandbarrettimages.co.uk/productimages/HB/70/048774_A.jpg',
        resolution: 70,
      },
      {
        path: '//images.hollandandbarrettimages.co.uk/productimages/HB/724/048774_A.jpg',
        resolution: 724,
      },
    ],
    group: '2',
  },
  {
    list: [
      {
        path: '//images.hollandandbarrettimages.co.uk/productimages/HB/370/048774_B.jpg',
        resolution: 370,
      },
      {
        path: '//images.hollandandbarrettimages.co.uk/productimages/HB/70/048774_B.jpg',
        resolution: 70,
      },
      {
        path: '//images.hollandandbarrettimages.co.uk/productimages/HB/724/048774_B.jpg',
        resolution: 724,
      },
    ],
    group: '3',
  },
]

describe('<ProductImage />', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<ProductImage id="048774" images={images} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('renders correctly with thumbnails', () => {
    const tree = renderer
      .create(<ProductImage id="048774" images={images} showThumbnails />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
