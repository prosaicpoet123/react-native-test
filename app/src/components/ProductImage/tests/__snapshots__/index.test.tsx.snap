// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductImage /> renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "https://images.hollandandbarrettimages.co.uk/productimages/HB/180/048774_A.jpg",
        }
      }
      style={
        Array [
          Object {
            "height": 180,
            "width": "100%",
          },
          Object {
            "height": 180,
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`<ProductImage /> renders correctly with thumbnails 1`] = `
<View>
  <View
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "https://images.hollandandbarrettimages.co.uk/productimages/HB/180/048774_A.jpg",
        }
      }
      style={
        Array [
          Object {
            "height": 180,
            "width": "100%",
          },
          Object {
            "height": 180,
          },
        ]
      }
    />
  </View>
  <View
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "flexWrap": "wrap",
        "marginVertical": 8,
      }
    }
  >
    <RCTScrollView
      data={
        Array [
          Object {
            "path": "//images.hollandandbarrettimages.co.uk/productimages/HB/70/048774_A.jpg",
            "resolution": 70,
          },
          Object {
            "path": "//images.hollandandbarrettimages.co.uk/productimages/HB/70/048774_B.jpg",
            "resolution": 70,
          },
        ]
      }
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={true}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
          style={
            Array [
              Object {
                "flexDirection": "row",
              },
              null,
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "borderColor": "#00594C",
                "borderWidth": 1,
                "opacity": 1,
              }
            }
          >
            <Image
              source={
                Object {
                  "uri": "https://images.hollandandbarrettimages.co.uk/productimages/HB/70/048774_A.jpg",
                }
              }
              style={
                Object {
                  "height": 70,
                  "width": 70,
                }
              }
            />
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={
            Array [
              Object {
                "flexDirection": "row",
              },
              null,
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Image
              source={
                Object {
                  "uri": "https://images.hollandandbarrettimages.co.uk/productimages/HB/70/048774_B.jpg",
                }
              }
              style={
                Object {
                  "height": 70,
                  "width": 70,
                }
              }
            />
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
